packages/lib/src/vis/tiles/TiledHeatmapMesh.tsx

```diff
@@ -1,71 +1,89 @@
 import { useThree } from '@react-three/fiber';
 import { clamp, range } from 'lodash';
+import { useRef } from 'react';
+import type { Group } from 'three';
 
 import { getInterpolator } from '../heatmap/utils';
 import { useCameraState } from '../hooks';
+import type { Size } from '../models';
 import { useAxisSystemContext } from '../shared/AxisSystemProvider';
 import TiledLayer from './TiledLayer';
 import type { TilesApi } from './api';
 import type { ColorMapProps } from './models';
-import { getScaledVisibleDomains } from './utils';
+import { getScaledVisibleBox } from './utils';
 
 interface Props extends ColorMapProps {
   api: TilesApi;
   displayLowerResolutions?: boolean;
   qualityFactor?: number;
+  size?: Size;
 }
 
 function TiledHeatmapMesh(props: Props) {
   const {
     api,
     displayLowerResolutions = true,
     qualityFactor = 1, // 0: Lower quality, less fetch; 1: Best quality
+    size,
     ...colorMapProps
   } = props;
   const { baseLayerIndex, baseLayerSize } = api;
 
   const canvasSize = useThree((state) => state.size);
   const { visSize } = useAxisSystemContext();
+  const meshSize = size ?? visSize;
 
-  const { xVisibleDomain, yVisibleDomain } = useCameraState(
-    (...args) => getScaledVisibleDomains(...args, baseLayerSize),
-    [baseLayerSize]
-  );
+  const groupRef = useRef<Group>(null);
 
-  const itemsPerPixel = Math.max(
-    1,
-    Math.abs(xVisibleDomain[1] - xVisibleDomain[0]) / canvasSize.width,
-    Math.abs(yVisibleDomain[1] - yVisibleDomain[0]) / canvasSize.height
+  const box = useCameraState(
+    (...args) =>
+      getScaledVisibleBox(...args, meshSize, baseLayerSize, groupRef),
+    [meshSize, baseLayerSize, groupRef]
   );
 
-  const roundingOffset = 1 - clamp(qualityFactor, 0, 1);
-  const subsamplingLevel = Math.min(
-    Math.floor(Math.log2(itemsPerPixel) + roundingOffset),
-    baseLayerIndex
-  );
-  const currentLayerIndex = baseLayerIndex - subsamplingLevel;
+  let layers: number[] = [];
+  if (box) {
+    const itemsPerPixel = Math.max(
```

Comment:
Seeing this variable name again, I wonder if `dataPointsPerPixel` would be clearer?